aci/resource_aci_fabricnodeidentp.go

```diff
@@ -263,35 +316,112 @@ func resourceAciFabricNodeMemberRead(ctx context.Context, d *schema.ResourceData
 	log.Printf("[DEBUG] %s: Beginning Read", d.Id())
 
 	aciClient := m.(*client.Client)
-
 	dn := d.Id()
 	fabricNodeIdentP, err := getRemoteFabricNodeMember(aciClient, dn)
-
 	if err != nil {
 		return errorForObjectNotFound(err, dn, d)
 	}
+
+	getAndsetDecommissionedNodes(aciClient, d.Get("serial").(string), d)
+
 	_, err = setFabricNodeMemberAttributes(fabricNodeIdentP, d)
 	if err != nil {
 		d.SetId("")
 		return nil
 	}
+
 	log.Printf("[DEBUG] %s: Read finished successfully", d.Id())
 
 	return nil
 }
 
 func resourceAciFabricNodeMemberDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
 	log.Printf("[DEBUG] %s: Beginning Destroy", d.Id())
-
 	aciClient := m.(*client.Client)
 	dn := d.Id()
-	err := aciClient.DeleteByDn(dn, "fabricNodeIdentP")
+
+	attachedToSwitch, err := verifyNodeAttachedToSwitch(aciClient, d.Get("serial").(string))
 	if err != nil {
 		return diag.FromErr(err)
 	}
+	if attachedToSwitch {
+		decommissionFabricNodeMemberFromController("true", "no", d, m)
+	} else {
+		err = aciClient.DeleteByDn(dn, "fabricNodeIdentP")
+		if err != nil {
+			return diag.FromErr(err)
+		}
+	}
 
 	log.Printf("[DEBUG] %s: Destroy finished successfully", d.Id())
 
 	d.SetId("")
 	return diag.FromErr(err)
 }
+func verifyNodeAttachedToSwitch(aciClient *client.Client, serial string) (bool, error) {
+	switchStatusDn := fmt.Sprintf("client-[%s]", serial)
+	dhcpClientCont, err := aciClient.Get(switchStatusDn)
+	if err != nil {
+		if !strings.Contains(err.Error(), "Error retrieving Object: Object may not exist") {
+			return false, err
+		}
+	}
+	data, _ := dhcpClientCont.ArrayElement(0, "imdata")
+	ip := G(data.S("dhcpClient", "attributes"), "ip")
+	nodeId := G(data.S("dhcpClient", "attributes"), "nodeId")
+	nodeRole := G(data.S("dhcpClient", "attributes"), "nodeRole")
+	if (ip != "0.0.0.0" && nodeId != "0") && (nodeRole == "leaf" || nodeRole == "spine") {
+		return true, nil
+	}
+	return false, nil
+}
+
+func decommissionFabricNodeMemberFromController(removeFromController string, commission string, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	log.Printf("[DEBUG] %s: Beginning removal of fabric node from controller", d.Id())
```

Comment:
Why the different end log compared to the this. Where here it is removal from fabric, and then at the end of the function it is a toggle between commission/decommission? Should we not keep them similar?

Reply:
changed